src/core/border_router/routing_manager.cpp

```diff
@@ -757,8 +757,17 @@ void RoutingManager::EvaluateRoutingPolicy(void)
     }
 
     // 3. Update advertised on-link & OMR prefixes information.
-    mIsAdvertisingLocalOnLinkPrefix = (newOnLinkPrefix == &mLocalOnLinkPrefix);
-    mAdvertisedOmrPrefixes          = newOmrPrefixes;
+    {
+        bool wasAdvertisingLocalOnLinkPrefix = mIsAdvertisingLocalOnLinkPrefix;
+
+        mIsAdvertisingLocalOnLinkPrefix = (newOnLinkPrefix == &mLocalOnLinkPrefix);
+        if (!wasAdvertisingLocalOnLinkPrefix && mIsAdvertisingLocalOnLinkPrefix)
+        {
+            InvalidateDiscoveredPrefixes();
+        }
+
+        mAdvertisedOmrPrefixes = newOmrPrefixes;
```

Comment:
I think here we try to  remove/invalidate the local prefix from `mDiscoveredPrefixes` (if it happens to be included) as soon as  we start advertising on-link prefix.

Can we just rely on the new logic in `InvalidateDiscoveredPrefixes()` to take care of this the next time it is called? I am not sure if there is any harm for it to stay in the list for a bit longer. Thoughts?

Reply:
Currently in `RoutingManager::UpdateDiscoveredOnLinkPrefix`, we don't add the local on-link prefix from PIO to `mDiscoveredPrefixes ` if `mIsAdvertisingLocalOnLinkPrefix = true`. 

It seems easier to comprehend if the following invariants are always true:
- When `mIsAdvertisingLocalOnLinkPrefix` is true, the local on-link prefix must not be in `mDiscoveredPrefixes`
- When `mIsAdvertisingLocalOnLinkPrefix` is false, the local on-link prefix may or may not be in `mDiscoveredPrefixes`

Reply:
SGTM.